doc/sphinx/Pacemaker_Development/components.rst

```diff
@@ -5,6 +5,80 @@ The Pacemaker code can be intricate and difficult to follow. This chapter has
 some high-level descriptions of how individual components work.
 
 
+.. index::
+   single: controller
+   single: pacemaker-controld
+
+Controller
+##########
+
+``pacemaker-controld`` is the Pacemaker daemon that utilizes the other daemons
+to orchestrate actions that need to be taken in the cluster. It receives CIB
+change notifications from the CIB manager, passes the new CIB to the scheduler
+to determine whether anything needs to be done, uses the executor and fencer to
+execute any actions required, and sets failure counts (among other things) via
+the attribute manager.
+
+As might be expected, it has the most code of any of the daemons.
+
+.. index::
+   single: join
+
+Join sequence
+_____________
+
+Most daemons track their cluster peers using Corosync's membership and CPG
+only. The controller additionally makes peers run a gauntlet referred to as
+"joining" (also the join sequence or join process). The join sequence allows
+the DC to ensure nodes are ready to be assigned tasks.
+
+A node's current join phase is tracked by the ``join`` member of ``crm_node_t``
+(used in the peer cache). It is an ``enum crm_join_phase`` that (ideally)
+progresses from the DC's point of view as follows:
+
+* The node initially starts at ``crm_join_none``
+* Once the DC sends the node a join offer, it proceeds to
+  ``crm_join_welcomed``.
+
+  This can happen in two ways: the DC's peer status callback notices when the
+  node joins the messaging layer (which registers ``I_NODE_JOIN`` ->
+  ``A_DC_JOIN_OFFER_ONE`` -> ``do_dc_join_offer_one()`` ->
+  ``join_make_offer()`` to unconfirmed nodes); and/or the joining node sends a
+  join announce at its controller startup.
+
+  The two may overlap. The DC can send a join offer and the node can send a
+  join announce at nearly the same time, so the node responds to the join offer
+  with a join request, and the DC responds to the join announce with a new join
+  offer. The process is designed to resolve itself but it can loop a bit.
```

Comment:
I would like to define (or mention) "join request" (see next paragraph) before we use it in this context. At first I wondered, "is a join request the same thing as a join announce?" while playing out the scenario.

Reply:
Makes sense